src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java

```diff
@@ -260,4 +245,90 @@ private void checkByteOrder(ValueLayout vl) {
             throw new IllegalArgumentException("Layout does not have the right byte order: " + vl);
         }
     }
+
+    private static int firstVariadicArgIndex(Option[] options) {
+        for (Option option : options) {
+            if (option instanceof LinkerOptions.FirstVariadicArg fva) {
+                return fva.index();
+            }
+        }
+        return -1;
+    }
+
+    // C spec mandates that variadic arguments smaller than int are promoted to int,
+    // and float is promoted to double
+    // See: https://en.cppreference.com/w/c/language/conversion#Default_argument_promotions
+    private static FunctionDescriptor promoteVariadicArgs(FunctionDescriptor function, int firstVariadicArgIndex) {
```

Comment:
Maybe it's unneeded optimization, but we're scanning the list of layouts once, and then once again to determine the resulting method type. Should we try to do both in one pass?

Reply:
I considered keeping track of the promoted args in a boolean[], but I figured it was simpler to just scan the layouts again. And this is not really performance critical code.